source/compiler/qsc_qasm/src/types.rs

```diff
@@ -59,74 +59,57 @@ impl Display for CallableKind {
     }
 }
 
-/// QASM supports up to seven dimensions, but we are going to limit it to three.
-#[derive(Debug, Clone, PartialEq, Eq)]
+/// QASM supports up to seven dimensions.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
 pub enum ArrayDimensions {
-    One(usize),
-    Two(usize, usize),
-    Three(usize, usize, usize),
-    Four(usize, usize, usize, usize),
-    Five(usize, usize, usize, usize, usize),
-    Six(usize, usize, usize, usize, usize, usize),
-    Seven(usize, usize, usize, usize, usize, usize, usize),
+    One,
+    Two,
+    Three,
+    Four,
+    Five,
+    Six,
+    Seven,
+}
+
+impl From<ArrayDimensions> for u32 {
+    fn from(value: ArrayDimensions) -> Self {
+        match value {
+            ArrayDimensions::One => 1,
+            ArrayDimensions::Two => 2,
+            ArrayDimensions::Three => 3,
+            ArrayDimensions::Four => 4,
+            ArrayDimensions::Five => 5,
+            ArrayDimensions::Six => 6,
+            ArrayDimensions::Seven => 7,
+        }
+    }
+}
```

Comment:
This can be a little more tersely written as:

```suggestion
pub enum ArrayDimensions {
    One = 1,
    Two = 2,
    Three = 3,
    Four = 4,
    Five = 5,
    Six = 6,
    Seven = 7,
}

impl From<ArrayDimensions> for u32 {
    fn from(value: ArrayDimensions) -> Self {
        value as u32
    }
}
```